packages/apps/webhooks/src/events-to-webhooks-app.ts

```diff
@@ -0,0 +1,183 @@
+      && (!doc.financial_status || doc.financial_status.current === 'pending')
+    ) {
+      return null;
+    }
+
+    logger.log(`> Sending ${isCart ? 'cart' : 'order'} notification`);
+
+    const token = options.webhook_token;
+    let headers: { Authorization: string; } | undefined;
+
+    if (token) {
+      headers = {
+        Authorization: `Bearer ${token}`,
+      };
+    }
+    const body = {
+      [isCart ? 'cart' : 'order']: doc,
+      customer,
+    };
+
+    try {
+      const { status } = await axios.post(url, body, { headers });
+      if (status) {
+        await updateManualQueue(manualQueue, appId);
+        // logger.log(`> ${status}`);
+        return status;
+      }
+    } catch (error: any) {
+      if (error.response && error.config) {
+        const err: { [x: string]: any } = { name: `#${docId} POST to ${error.config.url} failed` };
+        const { status, data } = error.response;
+        err.response = {
+          status,
+          data: JSON.stringify(data),
+        };
+        err.data = JSON.stringify(error.config.data);
+        logger.error(err);
+      } else {
+        logger.error(error);
+      }
+    }
+  }
+  return null;
+};
+
+const handleApiEvent: ApiEventHandler = async ({
+  evName,
+  apiEvent,
+  apiDoc,
+  app,
+}) => {
+  const appData = { ...app.data, ...app.hidden_data };
+  const resourceId = apiEvent.resource_id;
+  const key = `${evName}_${resourceId}`;
+
+  if (
+    Array.isArray(appData.ignore_events)
+    && appData.ignore_events.includes(evName)
+  ) {
+    logger.info('>> ', key, ' - Ignored event');
+    return null;
+  }
+
+  try {
+    if (appData.webhooks) {
+      let docId: string | undefined;
+      let manualQueue: any[] | null = null;
+      let isCart: boolean | undefined;
+
+      if (evName === 'applications-dataSet') {
+        const body = appData;
+        if (body && Array.isArray(body.manual_queue) && body.manual_queue.length) {
+          manualQueue = body.manual_queue;
+          const nextId = manualQueue[0];
+          if (typeof nextId === 'string' && /[a-f0-9]{24}/.test(nextId)) {
+            docId = nextId.trim();
+          }
+          manualQueue.shift();
+        }
+      } else {
+        isCart = evName === 'carts-customerSet' || evName === 'carts-new';
+        docId = apiDoc._id;
+      }
+
+      if (docId) {
+        let customer: Customers | undefined;
+        let doc: Carts | Orders;
+        if (isCart) {
+          doc = apiDoc as Carts;
+          if (doc.completed || doc.available === false) {
+            return null;
+          }
+          const abandonedCartDelay = 3 * 1000 * 60;
+          if (Date.now() - new Date(doc.created_at).getTime() >= abandonedCartDelay) {
+            const customerId = doc.customers && apiDoc.customers[0];
+            if (customerId) {
+              customer = (await api.get(`customers/${customerId}`)).data;
+            }
+          } else {
+            return null;
```

Comment:
https://github.com/ecomplus/app-webhooks/blob/master/functions/routes/ecom/webhook.js#L78

Tá indo de `return res.sendStatus(501)` pra `return null`, na prática na implementação na v1 tem retry, na sua aqui não teria,  nunca trataria carrinho nenhum na verdade.

Reply:
Aproveita e relembra aí se fez isso em algum outro lugar 🙃

Reply:
Vamos ter que arrumar outra maneira de tratar delay de carrinho abandonado aqui, não pode ser por retentativas da função porque nesse caso teriam execuções demais (o delay é curto) e além disso nós já ignoramos eventos com mais de 1min por padrão https://github.com/ecomplus/cloud-commerce/blob/main/packages/firebase/src/helpers/pubsub.ts#L26-L29

Reply:
Salvar no firestore ( erros >= 500), talvez com ID, o tipo (se é um carrinho ou pedido) e quantidade de retentativas já executadas (para limitar por x retentativas) e ter uma função agendada que irá buscar no firestore as retentativas.

Reply:
Ou dessa forma poderiamos causar outros problemas?

Reply:
> Salvar no firestore ( erros >= 500)

Só pontuando, na real não é todo "antigo" erro 5xx o problema aqui, é só com carrinho abandonado mesmo, que simulava um erro 5xx pra forçar rententativas até dar o delay pra considerar o carrinho como abandonado mesmo. Na grande maioria dos outros erros 5xx é só retornar erro normalmente pra deixar a retentativa normal do Cloud Function mesmo...

> quantidade de retentativas já executadas (para limitar por x retentativas)

Na verdade já existe um limite de retentativas, não é esse o problema. O negócio do carrinho é que essa retentativa (uma só) deveria ser agendada com um delay maior (alguns minutos, não segundos).

> função agendada que irá buscar no firestore as retentativas

Funciona, é a ideia mais bate-pronto, mas "função agendada" é o que eu sempre vou querer evitar 🙃

Reply:
O ideal seria conseguir criar um evento específico `carts-abandoned` e conseguir pegar já aqui https://github.com/ecomplus/cloud-commerce/blob/main/packages/firebase/src/handlers/check-store-events.ts somente os carrinhos já com esse tempo pós criação, aí não tem retentativa nenhuma.
Na prática essa rententativa forçada com 501 nas implementações atuais é uma gambiarra.